### ISSUE TITLE
[iOS] Location Authorization "Allow Once" on iOS 13 fails on 2nd run

### ISSUE THREAD

**Issue by ucsbricks** [dpa99c/cordova-diagnostic-plugin#367]
**I'm submitting a ...**
- [X] bug report
- [ ] feature request
- [ ] documentation issue

# Bug report

**Current behavior:**

On iOS 13.x user options on location permission dialog have been changed by Apple. If the user choses the newly added option "Allow Once", when cordova.plugins.diagnostic.requestLocationAuthorization() is called, result equals to cordova.plugins.diagnostic.permissionStatus.GRANTED_WHEN_IN_USE on first run of the app (which is correct). If one quits the app and restarts it, cordova.plugins.diagnostic.requestLocationAuthorization() returns no result at all. Neither success nor error callback is triggered.

**Expected behavior:**

As outlined [here](https://medium.com/q42-engineering/apple-location-permission-ios13-1e0e59002889), I would expect that if the user chose "Allow Once", result equals to cordova.plugins.diagnostic.permissionStatus.GRANTED_WHEN_IN_USE, _but_ authorization status is NOT_DETERMINED on the _next_ run of the app. Therefore calling cordova.plugins.diagnostic.requestLocationAuthorization() would trigger a permission dialog, exactly as on the first run of the app.

**Steps to reproduce:**

- Call cordova.plugins.diagnostic.requestLocationAuthorization()
- Choose option "Allow Once" (result is cordova.plugins.diagnostic.permissionStatus.GRANTED_WHEN_IN_USE)
- Force quit then restart the app
- Call cordova.plugins.diagnostic.requestLocationAuthorization() (there is no result or error)

**Environment information**

- Cordova CLI version 
	- 8.1.2 (cordova-lib@8.1.1)
- Cordova platform version
	- Installed platforms:
		- android 7.1.1
		- ios 4.5.5
- Plugins & versions installed in project (including this plugin)
	- com.ozexpert.devicemeta 0.0.2 "DeviceMeta"
	- com.rjfun.cordova.httpd 0.9.2 "CorHttpd"
	- cordova-custom-config 5.0.3 "cordova-custom-config"
	- cordova-labs-local-webserver 2.4.0 "CordovaLocalWebServer"
	- cordova-plugin-advanced-http 2.0.9 "Advanced HTTP plugin"
	- cordova-plugin-android-permissions 1.0.0 "Permissions"
	- cordova-plugin-app-version 0.1.9 "AppVersion"
	- cordova-plugin-appminimize 1.0.1 "AppMinimize"
	- cordova-plugin-apprate 1.4.0 "AppRate"
	- cordova-plugin-device 2.0.4-dev "Device"
	- cordova-plugin-dialogs 2.0.1 "Notification"
	- cordova-plugin-email-composer 0.8.3 "EmailComposer"
	- cordova-plugin-file 6.0.1 "File"
	- cordova-plugin-file-transfer 1.7.1 "File Transfer"
	- cordova-plugin-geolocation 4.0.1 "Geolocation"
	- cordova-plugin-globalization 1.11.0 "Globalization"
	- cordova-plugin-inappbrowser 3.0.0 "InAppBrowser"
	- cordova-plugin-inapppurchase 1.2.0 "In App Purchase"
	- cordova-plugin-localization-strings 3.2.0 "Localization"
	- cordova-plugin-nativestorage 2.3.2 "NativeStorage"
	- cordova-plugin-network-information 2.0.3-dev "Network Information"
	- cordova-plugin-networkactivityindicator 0.1.1 "NetworkActivityIndicator"
	- cordova-plugin-safariviewcontroller 1.5.4 "SafariViewController"
	- cordova-plugin-screen-orientation 3.0.1 "Screen Orientation"
	- cordova-plugin-splashscreen 5.0.2 "Splashscreen"
	- cordova-plugin-statusbar 2.4.2 "StatusBar"
	- cordova-plugin-whitelist 1.3.3 "Whitelist"
	- cordova-plugin-wkwebview-engine 1.1.5-dev "Cordova WKWebView Engine"
	- cordova-plugin-x-socialsharing 5.4.4 "SocialSharing"
	- cordova-plugin-zip 3.1.0 "cordova-plugin-zip"
	- cordova-support-google-services 1.1.0 "cordova-support-google-services"
	- cordova.plugins.diagnostic 4.0.12 "Diagnostic"
	- es6-promise-plugin 4.2.2 "Promise"
	- phonegap-plugin-mobile-accessibility 1.0.5-dev "Mobile Accessibility"
	- phonegap-plugin-multidex 1.0.0 "Multidex"
	- phonegap-plugin-push 2.2.3 "PushPlugin"
- Dev machine OS and version, e.g.
    - OSX
        - ProductName:    Mac OS X
        - ProductVersion: 10.14.6
        - BuildVersion:   18G95
        
_Runtime issue_
- Device details
    - iPad 6th Gen, iPad Pro 9.7", iPhone 8
- OS details
    - iOS 13.1

**Comment by dpa99c** [comment]
I'm not able to reproduce this behaviour when running the [example project app](https://github.com/dpa99c/cordova-diagnostic-plugin-example) on an iPhone running iOS 13.1 - the behaviour I observed is as would be expected:
- Calling `cordova.plugins.diagnostic.requestLocationAuthorization()` and choosing option "Allow Once" results in `cordova.plugins.diagnostic.permissionStatus.GRANTED_WHEN_IN_USE` and the app can access location
- Forcing the app to quit then restarting the app results in a status of `NOT_DETERMINED`
- Calling `cordova.plugins.diagnostic.requestLocationAuthorization()` again shows the permissions dialog again.

Please try building and running the example project app to confirm whether you observe the same result in the same known app environment.

**Comment by fcamblor** [comment]
I'm facing a similar issue with this new "Allow once" status.

I suspect that the outcome of `requestLocationAuthorization()` call is based on the "button index" of the clicked option rather than its true value.

Let me elaborate : when I execute this code snippet as the very first execution of my app (just after installation) :
```
cordova.plugins.diagnostic.requestLocationAuthorization(console.info, console.error, cordova.plugins.diagnostic.locationAuthorizationMode.ALWAYS);
```

I get following outcome :
- Popup is opened with *only 3* options : When in use (selected by default) / Allow once / Never (the "Always" option is not shown anymore since iOS 13 !)
- When first option (When in use) is selected, the outcome of the call is an "authorized" (= cordova.plugins.diagnostic.permissionStatus.GRANTED constant) value
- When second option (Allow once) is selected, the outcome of the call is an "authorized_when_in_use" (= cordova.plugins.diagnostic.permissionStatus.GRANTED_WHEN_IN_USE constant) value
- When third option (Never) is selected, the outcomt of the call is a "denied_always" (= cordova.plugins.diagnostic.permissionStatus.DENIED_ALWAYS constant) value

Previously, on iOS < 13, first option was "Always" and second one was "When in use" ... that's why I suspect that the outcome is based on the index of the selected option rather than the true selected value.

When I open the System preferences and navigate to my app preferences, I can see that what I selected in the popup is the truth, and `diagnostic` plugin saved value is not the good one.

Note : I tested with `cordova.plugins.diagnostic.locationAuthorizationMode.WHEN_IN_USE` third `mode` parameter and it doesn't change anything to my ascertainment.

TLDR; I have 2 major concerns regarding `requestLocationAuthorization` with iOS 13 :

1. its outcome is not true currently if I either select "When in use" or "Allow once" options (the "Never" one is still 👌 )
2. I don't have the possibility to select "Always" option anymore from within my app. I'm not sure if you can do something for this in the plugin (allow to display the 4th "Always" option in the permission popup) but if you can't it would be a real pain to ask users to open their iOS preferences to switch to this mode as this is a very complicated manipulation :/

**Comment by fcamblor** [comment]
After reading https://medium.com/q42-engineering/apple-location-permission-ios13-1e0e59002889 it seems like some of my assumptions were not correct :
- It seems like "Allow once" should be considered exactly like the "When in use" outcome :
> If the user taps option 1: “Allow once”, your app will be notified that the CLAuthorizationStatus changed to authorizedWhenInUse. Just like you’re used to in older iOS version when you get a permanent permission. It is now allowed for your app to start requesting locations, no code changes necessary.
- Background acquisition of geoposition will then be asked to the user on purpose, only when there is a background process that acquires those positions
> Getting consent to scan for location in the background has gotten a bit harder but in exchange, your users will enjoy more transparency regarding the locations you’ve received in iOS 13. The first notable change is that even when you call `requestAlwaysAuthorization`, the user will only get the ‘just now’ and ‘when in use’ options in the permission dialog. If the user grants you ‘when in use’ permission and you try to scan for location in the background, only then the user will be presented a dialog to grant the background permission.

So ... not sure what to think about it, it looks like the current implementation of the plugin is right.
But I find it misleading that when you select "When in use", you get a "GRANTED" outcome.
I can understand it from a backward compatibility view, but it sounds really misleading to me.

@dpa99c I would be really interested in your thoughts on this :-)

**Comment by dpa99c** [comment]
@fcamblor As you can see from [the source code](https://github.com/dpa99c/cordova-diagnostic-plugin/blob/master/src/ios/Diagnostic_Location.m#L119), this plugin is simply translating the [native iOS location authorization constants](https://developer.apple.com/documentation/corelocation/clauthorizationstatus?language=objc) to string values in order to pass them back up to the JS layer in a Cordova app.

iOS 13 has made some drastic (and questionable) changes to how location permissions work. This plugin is simply a thin wrapper around the native APIs that allow requesting and querying of location authorization on iOS. However, the CoreLocation APIs have not changed in iOS 13 and therefore there's nothing this plugin can do programmatically to handle the change in behaviour which iOS 13 has introduced. Therefore it is up to the app using this plugin to handle then new behaviour.

As far as the labelling of the constants - `GRANTED` is a hangover from iOS 8 when there was no distinction between `WHEN_IN_USE` and `ALWAYS` however changing it to the latter makes it trickier to handle cross-platform differences since the current Android implementation currently handles only `GRANTED` vs `DENIED`.
Having said that, the Android 10 (Q) SDK introduces the concept of background vs foreground location permission so when the Android implementation is updated to handle that, that may be the time to update the iOS implementation also to relabel the constant `GRANTED` as `GRANTED_ALWAYS`.

However, I'm closing this for now as I don't believe there's anything that can be done programmatically to handle the quirks of requesting location introduced in iOS 13.
